### PR TITLE
feat: handle implicit args in runner

### LINKED ISSUE #255: dev: update test runner to accept args for implicits args

### LINKED ISSUE #256: dev: update test runner to return implicits

### PR DESCRIPTION
**Pull request by enitrat**
Closes #255
Closes #256
trust me, it works

### CODE REVIEW

File: cairo/tests/fixtures/runner.py [hunk 1]
```diff
@@ -58,13 +59,40 @@ def cairo_run(request, cairo_program, cairo_file, main_path):
...
+            ScopedName(path=("__main__", entrypoint, "ImplicitArgs"))
+        ).members.items()
+
+        # Split implicit args into builtins and other implicit args
+        _builtins = {
+            k: {
+                "python_type": to_python_type(
+                    resolve_main_path(main_path)(v.cairo_type)
+                ),
+                "cairo_type": v.cairo_type,
```
Comment (enitrat) [inline cairo/tests/fixtures/runner.py:72]: we will need both to return implicit args later.

File: cairo/tests/fixtures/runner.py [hunk 2]
```diff
@@ -166,7 +192,8 @@ def _factory(entrypoint, *args, **kwargs):
 
         runner.end_run(disable_trace_padding=False)
         if request.config.getoption("proof_mode"):
-            return_data_offset = serde.get_offset(return_data.cairo_type)
+            # TODO: should we also support multiple return data types in proof mode?
```
Comment (enitrat) [inline cairo/tests/fixtures/runner.py:195]: @ClementWalter not sure I understand what should be done in proof mode here. do I want the implicits to be included?

File: cairo/tests/fixtures/runner.py [hunk 3]
```diff
@@ -58,39 +59,56 @@ def cairo_run(request, cairo_program, cairo_file, main_path):
...
+        _implicit_args = {
+            k: {
+                "python_type": to_python_type(
+                    resolve_main_path(main_path)(v.cairo_type)
+                ),
+                "cairo_type": v.cairo_type,
+            }
+            for k, v in implicit_args
+            if not any(builtin in k.replace("_ptr", "") for builtin in ALL_BUILTINS)
+        }
```
Comment (ClementWalter) [inline cairo/tests/fixtures/runner.py:87]:
```suggestion
        implicit_args = cairo_program.identifiers.get_by_full_name(
            ScopedName(path=("__main__", entrypoint, "ImplicitArgs"))
        ).members

        # Split implicit args into builtins and other implicit args
        _builtins = {
            k.replace("_ptr", "")
            for k in implicit_args.values()
            if any(builtin in k.replace("_ptr", "") for builtin in ALL_BUILTINS)
        }

        _implicit_args = {
            k: {
                "python_type": to_python_type(
                    resolve_main_path(main_path)(v.cairo_type)
                ),
                "cairo_type": v.cairo_type,
            }
            for k, v in implicit_args.items()
            if not any(builtin in k.replace("_ptr", "") for builtin in ALL_BUILTINS)
        }
```

File: cairo/tests/fixtures/runner.py [hunk 4]
```diff
@@ -58,39 +59,56 @@ def cairo_run(request, cairo_program, cairo_file, main_path):
...
-                "ec_op",
-                "keccak",
-                "poseidon",
-                "range_check96",
-                "add_mod",
-                "mul_mod",
-            ]
-            if builtin in {arg.replace("_ptr", "") for arg in implicit_args}
+            for builtin in ALL_BUILTINS
+            if builtin in {arg.replace("_ptr", "") for arg in _builtins.keys()}
```
Comment (ClementWalter) [inline cairo/tests/fixtures/runner.py:112]: can probably replace with a set intersect here

File: cairo/tests/fixtures/runner.py (hunk 2 above)
Comment (ClementWalter) [reply cairo/tests/fixtures/runner.py:196]: proof mode needs this to include returned data into the public memory, implicit args are not meant to be different but, if only builtins, were done in the builtin final stack

File: cairo/tests/fixtures/runner.py [hunk 5]
```diff
@@ -255,20 +282,24 @@ def _factory(entrypoint, *args, **kwargs):
...
-                else [function_output]
+
+        function_output = []
+        return_data_offsets = serde.get_offsets(return_data_types)
+        for i, return_data_type in enumerate(return_data_types):
+            function_output.append(
+                serde.serialize(
+                    return_data_type, runner.vm.run_context.ap, return_data_offsets[i]
+                )
             )
```
Comment (ClementWalter) [inline cairo/tests/fixtures/runner.py]:
```suggestion
        return_data_offsets = serde.get_offsets(return_data_types)
        function_output = [
            serde.serialize(
                return_data_type, runner.vm.run_context.ap, offset
            )
            for offset, return_data_type in zip(return_data_offsets, return_data_types)            
        ]
```

File: cairo/tests/utils/serde.py [hunk 6]
```diff
@@ -286,6 +297,16 @@ def get_offset(self, cairo_type):
...
                 return 1
 
+    def get_offsets(self, cairo_types: List[CairoType]):
+        """Given a list of Cairo types, return the offset for each type."""
+        offsets = []
+        total = 0
+        for cairo_type in reversed(cairo_types):
+            offset = self.get_offset(cairo_type)
+            total += offset
+            offsets.append(total)
```
Comment (ClementWalter) [inline cairo/tests/utils/serde.py:307]: this is a cumsum over the offset, make this explicit and pythonic, see https://stackoverflow.com/questions/15889131/how-to-find-the-cumulative-sum-of-numbers-in-a-list/15889203#15889203 (itertools or iteration with :=)

**Review by ClementWalter — CHANGES_REQUESTED** [review]
see comments

File: cairo/tests/fixtures/runner.py (hunk 4 above)
Comment (enitrat) [reply cairo/tests/fixtures/runner.py]: not going to use a set here as it changes the order of the builtins, but the cairo-lang library _requires_ them to be in the same order

File: cairo/tests/fixtures/runner.py (hunk 3 above)
Comment (enitrat) [reply cairo/tests/fixtures/runner.py]: not replacing the "ptr" in _builtins here with an empty string, as it further complexifies the process of replacing the _ptr suffix by `builtin` when handling builtins down the line

**Review by ClementWalter — APPROVED** [review]

**Pull request merged**